src/main/java/org/cbioportal/security/config/OAuth2SecurityConfig.java

```diff
@@ -1,110 +1,156 @@
     
-    @Bean
-    @ConditionalOnProperty(value = "authenticate", havingValue = "oauth2")
-    public SecurityFilterChain oAuth2filterChain(HttpSecurity http) throws Exception {
-
-        http.authorizeHttpRequests(auth -> 
-                auth.requestMatchers("/api/health", "/login", "/images/**").permitAll()
-                    .anyRequest().authenticated())
-            .oauth2Login(oauth -> oauth.loginPage("/login"))
-            .logout((logout) -> logout.logoutSuccessUrl("/login?logout_success"))
-            .exceptionHandling(eh ->
-                eh.defaultAuthenticationEntryPointFor(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED), AntPathRequestMatcher.antMatcher("/api/**")))
-            .csrf(AbstractHttpConfigurer::disable);
-        
-        if(!Objects.isNull(this.jwtResourceServerUri) && !this.jwtResourceServerUri.isEmpty()) {
-            http.oauth2ResourceServer(OAuth2ResourceServerConfigurer::jwt);
-        }
-        return http.build();
-    }
+    // TODO - add this to portal.properties.EXAMPLE
+    // TODO - discuss changing this to /logout (Spring Security default) with Aaron
+    @Value("${oauth2.logout.url:/j_spring_security_logout}")
+    private String logoutUrl;
     
+    // FIXME - we cannot rely on the idp being named 'keycloak' here. In principle
+    // it can be any name, and there can be multiple idp's configured.
+    @Value("${spring.security.oauth2.client.registration.keycloak.client-id:cbioportal_oauth2}")
+    private String clientId;
+
+    // TODO - implement configured retrieval of roles based on this property
+    @Value("${spring.security.oauth2.client.jwt-roles-path:resource_access::cbioportal::roles}")
+    private String jwtRolesPath;
+
+    @Autowired
+    private ClientRegistrationRepository clientRegistrationRepository;
+
     @Bean
-    @ConditionalOnProperty(value = "authenticate", havingValue = "optional_oauth2")
-    public SecurityFilterChain optionalOAuth2filterChain(HttpSecurity http) throws Exception {
-        return http
-            .oauth2Login(oauth -> oauth.loginPage("/login"))
-            .authorizeHttpRequests(auth -> auth
-                .requestMatchers("/").permitAll()
-                .anyRequest().permitAll())
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        http
+            // FIXME - csrf should be enabled
             .csrf(AbstractHttpConfigurer::disable)
-            .logout((logout) -> logout.logoutSuccessUrl("/"))
-            .build();
+            .authorizeHttpRequests(authorize -> 
+                authorize
+                    .requestMatchers("/api/health", "/login", "/images/**").permitAll()
+                    .anyRequest().authenticated()
+            )
+            .oauth2Login(login ->
+                login.userInfoEndpoint(userInfo ->
+                    userInfo.userAuthoritiesMapper(userAuthoritiesMapper())
+                )
+            )
+            .logout(logout -> logout
```

Comment:
I think that in this form of OAuth2 login (there is one upstream OIDC IDP) the logout should be transmitted to the IDP. AFAIK the logoutSuccessHandler is invoked after local logout by Spring Security. I included a link to more information on this in a comment:

 // See: https://docs.spring.io/spring-security/reference/5.7-SNAPSHOT/servlet/oauth2/login/advanced.html#oauth2login-advanced-oidc-logout

If I would not do this. The user would not be able to logout. The following would occur:
1. User presses logout button
2. cBioPortal screen reloads

This would be the result of:
1. Spring performing local logout
2. User is directed to /login
3. User is redicted to IDP
4. IDP still has active session for user and redirects user to cBioPortal (user does not need to enter username/password)
5. cBioPortal screen loads

So, because of the SSO session with the IDP not being terminated and there is a single IDP gives the user the impression that no actual logout was performed.

This behavior is not applicable when there are multiple OAuth2 IDP, as you have in your examples (Google, Microsoft). In this case the user is kicked back to the IDP selection window when visiting /login.

I also do not know how to best solve this. I think there is different behavior needed when having a single or when having multiple OAuth2 IDPs.